In [1]:
import numpy as np
import pandas as pd
import scipy.sparse as sps

In [11]:
types = pd.read_csv('../Input/data_ICM_type.csv')

In [12]:
types

item_id  feature_id  data
0            0           1     1
1            1           3     1
2            2           4     1
3            3           1     1
4            4           3     1
...        ...         ...   ...
23086    27963           1     1
23087    27964           2     1
23088    27965           1     1
23089    27966           1     1
23090    27967           4     1

[23091 rows x 3 columns]

In [13]:
types = types.drop(columns = ['data'], axis=1)
types

item_id  feature_id
0            0           1
1            1           3
2            2           4
3            3           1
4            4           3
...        ...         ...
23086    27963           1
23087    27964           2
23088    27965           1
23089    27966           1
23090    27967           4

[23091 rows x 2 columns]

In [14]:
types = types.rename(columns={'feature_id':'type'})
types

item_id  type
0            0     1
1            1     3
2            2     4
3            3     1
4            4     3
...        ...   ...
23086    27963     1
23087    27964     2
23088    27965     1
23089    27966     1
23090    27967     4

[23091 rows x 2 columns]

In [16]:
typesFiltered = types[types['item_id'] <= 24506]

typesFiltered.head()

item_id  type
0        0     1
1        1     3
2        2     4
3        3     1
4        4     3

In [17]:
typesFiltered

item_id  type
0            0     1
1            1     3
2            2     4
3            3     1
4            4     3
...        ...   ...
19625    24502     1
19626    24503     4
19627    24504     1
19628    24505     4
19629    24506     1

[19630 rows x 2 columns]

In [38]:
typesArray = typesFiltered['type'].to_numpy()
typesArray

array([1, 3, 4, ..., 1, 4, 1])

In [27]:
itemsID = typesFiltered['item_id'].to_numpy()
itemsID

array([    0,     1,     2, ..., 24504, 24505, 24506])

In [15]:
URM = pd.read_csv('../Input/interactions_and_impressions.csv')
itemsURM = URM['ItemID'].to_numpy()
itemsURM = np.unique(itemsURM, axis=0)
itemsURM

/var/folders/tt/9v6608x550n51759gpcrffsc0000gn/T/ipykernel_47827/98915780.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  URM = pd.read_csv('../Input/interactions_and_impressions.csv')


array([    0,     1,     2, ..., 24504, 24505, 24506])

As we can the problem is that in URM we have 24507 items, but in the ICM we have info just about 23091.
Let's fill with 0 the items from which we don't have any info.

In [19]:
len(itemsURM)

24507

In [36]:
ICMType = np.zeros((len(itemsURM), 1))
ICMType

array([[0.],
       [0.],
       [0.],
       ...,
       [0.],
       [0.],
       [0.]])

In [37]:
for x in range(len(itemsID)):
    #print(itemsID[x])
    ICMType[itemsID[x]][0] = typesArray[x]
ICMType

array([[1.],
       [3.],
       [4.],
       ...,
       [1.],
       [4.],
       [1.]])

## Algoritmo per ICM

In [1]:
import numpy as np
import pandas as pd
import scipy.sparse as sps

## Types

In [2]:
types = pd.read_csv('../Input/data_ICM_type.csv')
types = types.drop(columns = ['data'], axis=1)
types = types.rename(columns={'feature_id':'type'})
types

item_id  type
0            0     1
1            1     3
2            2     4
3            3     1
4            4     3
...        ...   ...
23086    27963     1
23087    27964     2
23088    27965     1
23089    27966     1
23090    27967     4

[23091 rows x 2 columns]

In [3]:
typesFiltered = types[types['item_id'] <= 24506]
typesArray = typesFiltered['type'].to_numpy()
itemsID = typesFiltered['item_id'].to_numpy()

In [4]:
URM = pd.read_csv('../Input/interactions_and_impressions.csv')
itemsURM = URM['ItemID'].to_numpy()
itemsURM = np.unique(itemsURM, axis=0)
itemsURM

/var/folders/tt/9v6608x550n51759gpcrffsc0000gn/T/ipykernel_60746/98915780.py:1: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  URM = pd.read_csv('../Input/interactions_and_impressions.csv')


array([    0,     1,     2, ..., 24504, 24505, 24506])

In [30]:
ICM = np.zeros((len(itemsURM), 2), dtype=int)

In [31]:
for x in range(len(itemsID)):
    ICM[itemsID[x]][0] = typesArray[x]

## Length

In [32]:
lenght = pd.read_csv('../Input/data_ICM_length.csv')
lenght = lenght.drop(columns = ['feature_id'], axis=1)
lenght = lenght.rename(columns={'data':'numberOfEpisodes'})
lenght

item_id  numberOfEpisodes
0            0                 1
1            1                 1
2            2                21
3            3                 1
4            4                 1
...        ...               ...
23086    27963                 1
23087    27964                 1
23088    27965                 1
23089    27966                 1
23090    27967                 2

[23091 rows x 2 columns]

In [33]:
lenghtFiltered = lenght[lenght['item_id'] <= 24506]
lenghtArray = lenghtFiltered['numberOfEpisodes'].to_numpy()
itemsID = typesFiltered['item_id'].to_numpy()

In [34]:
for x in range(len(itemsID)):
    ICM[itemsID[x]][1] = lenghtArray[x]

In [39]:
ICM

array([[   1,    1],
       [   3,    1],
       [   4,   21],
       ...,
       [   1,    1],
       [   4, 1001],
       [   1,    1]])